Refactor HMC and warning system

aseyboldt opened the pull request:
We recently implemented a couple of improvements for NUTS, and most of them would also be applicable to HMC (and some also to metropolis, but this is a separate issue).
This is a first step toward using dual averaging in hmc. Proper divergence handling is still missing in hmc. It might be better to move some of this into `BaseHMC`, and just use it in both NUTS and HMC.

aseyboldt committed:
Move DualAveraging out of nuts

aseyboldt committed:
Use dual averaging in hmc

twiecki commented on pymc3/step_methods/hmc/hmc.py:
```diff
@@ -50,18 +55,40 @@ def __init__(self, vars=None, path_length=2., step_rand=unif, **kwargs):
+            probability across the trajectories are close to target_accept.
+            Higher values for target_accept lead to smaller step sizes.
+        gamma : float, default .05
+        k : float (.5,1) default .75
```
.5,1?

aseyboldt replied:
I just copied that from nuts, so if we change it here, we should also change it there.
k has to be in the open interval (0.5, 1).

twiecki replied:
I see, isn't `[0.5, 1]` the more common syntax?

aseyboldt replied:
That would imply a closed interval, including 0.5 and 1. I don't think that's allowed. I can just put it in words in the description, that would be leas confusing I think.

ColCarroll replied:
I think common syntax might be `]0.5, 1[`, which makes me twitch, but at least makes it clear we're not talking about a tuple.  also fine with using words!

junpenglao replied:
Please dont use `]0.5, 1[`... word is good

aseyboldt replied:
Words it is.
I think the ]0.5, 1[ syntax is more common in the US, and (0.5, 1) in europe.

aseyboldt committed:
Update doc of dual average parameters

twiecki commented:
`pymc3/step_methods/hmc/hmc.py:7: [W0611(unused-import), ] Unused floatX imported from pymc3.theanof`

aseyboldt committed:
Rewrite warning system

aseyboldt committed:
Add stats to hmc

aseyboldt commented on pymc3/sampling.py:
```diff
@@ -289,6 +296,9 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None,
         Options for traceplot. Example: live_plot_kwargs={'varnames': ['x']}
     discard_tuned_samples : bool
         Whether to discard posterior samples of the tune interval.
+    compute_stats : bool, default=True
```
Any ideas for a better name than `compute_stats`?

ColCarroll replied:
`store_stats`?

aseyboldt replied:
Hm. That might be confusing with the normal sampler stats.
Maybe `convergence_checks` or `convergence_stats` or the very long `compute_convergence_stats`?

springcoil replied:
I think `compute_convergence_stats` suits.

aseyboldt commented:
I ended up refactoring more than I planned:

The hmc related code is divided into three classes: `BaseHMC`, `HamiltonianMC` and `NUTS`. The first implements things that are common to both hmc and nuts. In this PR I moved dual averaging (now wrapped in a separate class, which might also be useful for metropolis) and mass matrix adaptation from `NUTS` to `BaseHMC`, so that `HamiltonianMC` can take advantage of it as well. The actual trajectory is computed and sampled in `NUTS._hamiltonian_step` and `HamiltonianMC._hamiltonian_step`, which is called by `BaseHMC.astep`. `BaseHMC.astep` handles divergences, adaptation, warnings, and some error checking.

I rewrote the whole warning system. Warnings during sampling are no longer python warnings (although we could still raise those if we want), but are stored as a `namedtuple`:
```python
"""
At the moment `kind` can be one of
* "divergence" (for hmc and nuts)
* "tuning-divergence"
* "bad-params" (for problematic sampler parameters)
* "convergence" (for indications that the chains did not converge, eg Rhat)
* "bad-acceptance"
* "treedepth" (for nuts)
"""
SamplerWarning = namedtuple(
    'SamplerWarning',
    "kind, message, level, step, exec_info, extra")
```
Several classes have a `warnings(strace)` method now, that generates these warnings. This is still missing for the mass matrix adaptation, because I'm not sure yet, how to diagnose problems there.

`level` is the same as the logging levels from the std logging module. `exec_info` stores an exception if there was one, and `extra` can store additional data related to that particular warning. For divergences it stores the integration state at the point of the divergence (which is different than the later accepted point, and should be useful for plotting.) If there is a very large number of divergences we stop storing that state at some point, to avoid memory exhaustion.

`pm.sample` collects the warnings, and attaches them to the trace in a report, that keeps track of the chain in which the warning occurred. This report (`trace.report`) also computes `gelman_rubin` and `effective_n` and generates global warnings if those don't look ok. The report object could probably use a lot of work (a html representation for the notebook would be cool). Right now it has only two public functions: `ok` and `raise_ok`. I'm not sure if or how we should expose the warnings to the users directly.

`pm.sample` then calls `trace.report._log_warnings`, which uses the std logging module to print the warnings. The global logging level and the level of the individual warnings decides if the warning is shown to the user.

I think we should think a bit more about how we want to present those warnings to the user (eg do we want to use the std logging module, do we only want to show one warning "please look at the report" etc) but I think this can wait until after this PR, as at this point we only need to change the `_log_warnings` method.

junpenglao commented on pymc3/step_methods/hmc/base_hmc.py:
```diff
@@ -59,10 +86,114 @@ def __init__(self, vars=None, scaling=None, step_scale=0.25, is_cov=False,
+        raise NotImplementedError("Abstract method")
+
+    def astep(self, q0):
+        """Perform a single NUTS iteration."""
```
Nitpick: "a single HMC iteration"

aseyboldt replied:
fixed

junpenglao commented on pymc3/backends/base.py:
```diff
@@ -447,10 +463,13 @@ def get_sampler_stats(self, varname, burn=0, thin=1, combine=True,
 
-    def _slice(self, idx):
+    def _slice(self, slice):
         """Return a new MultiTrace object sliced according to `idx`."""
```
according to `idx` --> according to `slice`

aseyboldt replied:
fixed

junpenglao commented on pymc3/backends/report.py:
```diff
@@ -0,0 +1,145 @@
+                'convergence', msg, 'info', None, None, gelman_rubin)
+            warnings.append(warn)
+
+        eff_min = min(val.min() for val in effective_n.values())
```
would it be more informative to raise warning if the effect sample size is lower than a certain percentage of the total sample? For example, raise warning if effect sample size is below 25% of the total number of sample.

aseyboldt replied:
Good idea. We could have both, too. One with level `info` if it is <25% and one with level `warn` if it is <200?

junpenglao replied:
I would do the other way around:
`warn` if it is <25% (high autocorrelation and poor mixing is likely an indication of modelling problem) 
and `info` if it is <200 (sometimes ppl sample only 200 for demo purpose and it would be annoying to see warning)

aseyboldt replied:
I guess that comes down to how much we trust `effective_n`. If we don't trust the numbers it returns, but only use it as an indication that the sampler doesn't work well, then I think I agree. But if we take the values from `effective_n` at face value, then a low percentage of effective samples isn't in itself a problem (ie should only be `info`). But a very low number of eff samples is a problem.
Say your samples have some autocorrelation, but you don't get any divergences, and gelman_rubin looks fine. Then I don't see a problem with just running the sampler for a very long time, until you get a lot of effective samples. This seems to me like a valid use-case, and we'd print warnings if people did this, even though what they are doing is ok. (at least I think it is, right?)
On the other hand, few effective samples are always a problem if you plan to do anything with your trace.
Maybe a middle way would be:
- `info` warning if `neff < 0.25 * draws`
- `warn` if `neff < 200` *and* `draws > 500`

That way we don't issue a low-neff warning if users *explicitly* asked for a low number of samples.

junpenglao replied:
Yeah that's a good point, I agree with this solution.

ColCarroll commented on pymc3/backends/report.py:
```diff
@@ -0,0 +1,145 @@
+logger = logging.getLogger('pymc3')
+
+
+SamplerWarning = namedtuple(
```
might make sense to have `SamplerWarning.kind` be an `Enum`, so that there is a little more reuse?

aseyboldt replied:
yes, fixed

junpenglao commented:
Oh, and final nitpick: It would be great to change `pm._log` to `_log` using `logging` everywhere else in the code base. Of course, we can also do it in a separate PR.

aseyboldt committed:
Change SamplerWarning.kind to enum

aseyboldt commented:
I agree that we should do that, but I already had to rebase this PR because of that at least 3 times...
Maybe a separate PR is better :-)

junpenglao commented:
LOL, agree.
Travis was complaining about unused import previously:
```
************* Module pymc3.sampling
pymc3/sampling.py:21: [W0611(unused-import), ] Unused SamplerWarning imported from pymc3.backends.report
************* Module pymc3.tests.test_step
pymc3/tests/test_step.py:14: [W0611(unused-import), ] Unused SamplingError imported from pymc3
```
Did you fix that?

aseyboldt commented:
ups, I have now :-)

junpenglao commented:
This is good to go.

springcoil commented:
Cool - excited about all of this :)

aseyboldt commented:
Ah, I thought we wanted to wait with this one until after 3.3 so that we have a bit more time to find problems with it? I guess we could also just branch off a 3.3 branch from before this merge, and push that to github. We want to release that pretty soon, right?

junpenglao commented:
Ohhh right. Well then definitely branch off 3.3 before this merge 
cc @twiecki @fonnesbeck

twiecki commented:
How about we just revert the merge commit?

junpenglao commented:
For the release? Sure. Sorry about the potential mess in the history...

junpenglao commented:
Re v3.3 release, maybe we can just include this PR (after merging #2808)?
